internal/controller/build/controller_conditions.go

```diff
@@ -0,0 +1,189 @@
+		ReasonAutoDeploymentFailed,
+		"Deployment configuration failed.",
+		generation,
+	)
+}
+
+func NewAutoDeploymentSuccessfulCondition(generation int64) metav1.Condition {
+	return controller.NewCondition(
+		ConditionDeploymentApplied,
+		metav1.ConditionTrue,
+		ReasonAutoDeploymentApplied,
+		"Successfully configured the deployment.",
+		generation,
+	)
+}
+
+func (r *Reconciler) markStepAsSucceeded(build *choreov1.Build, conditionType controller.ConditionType) {
+	successDescriptors := map[controller.ConditionType]struct {
+		Reason  controller.ConditionReason
+		Message string
+	}{
+		ConditionCloneSucceeded: {
+			Reason:  ReasonCloneSucceeded,
+			Message: "Source code cloning was successful.",
+		},
+		ConditionBuildSucceeded: {
+			Reason:  ReasonBuildSucceeded,
+			Message: "Building the source code was successful.",
+		},
+		ConditionPushSucceeded: {
+			Reason:  ReasonPushSucceeded,
+			Message: "Pushing the built image to the registry was successful.",
+		},
+	}
+
+	meta.SetStatusCondition(&build.Status.Conditions, controller.NewCondition(
+		conditionType,
+		metav1.ConditionTrue,
+		successDescriptors[conditionType].Reason,
+		successDescriptors[conditionType].Message,
+		build.Generation,
+	))
+}
+
+func (r *Reconciler) markStepAsFailed(build *choreov1.Build, conditionType controller.ConditionType) {
+	failureDescriptors := map[controller.ConditionType]struct {
+		Reason  controller.ConditionReason
+		Message string
+	}{
+		ConditionCloneSucceeded: {
+			Reason:  ReasonCloneFailed,
+			Message: "Source code cloning failed.",
+		},
+		ConditionBuildSucceeded: {
+			Reason:  ReasonBuildFailed,
+			Message: "Building the source code failed.",
+		},
+		ConditionPushSucceeded: {
+			Reason:  ReasonPushFailed,
+			Message: "Pushing the built image to the registry failed.",
+		},
+	}
+
+	meta.SetStatusCondition(&build.Status.Conditions, controller.NewCondition(
+		conditionType,
+		metav1.ConditionFalse,
+		failureDescriptors[conditionType].Reason,
+		failureDescriptors[conditionType].Message,
+		build.Generation,
+	))
+
+	meta.SetStatusCondition(&build.Status.Conditions, controller.NewCondition(
+		ConditionCompleted,
+		metav1.ConditionFalse,
+		ReasonWorkflowFailed,
+		"Build completed with a failure status",
+		build.Generation,
+	))
+}
+
+func (r *Reconciler) markWorkflowCompleted(build *choreov1.Build, argoPushStepOutput *argoproj.Outputs) {
+	newCondition := metav1.Condition{
+		Type:               string(ConditionCompleted),
+		Status:             metav1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+		Reason:             string(ReasonWorkflowCompleted),
+		Message:            "Build completed successfully.",
+	}
+	image := getImageNameFromWorkflow(*argoPushStepOutput)
+	if image == "" {
+		newCondition.Status = metav1.ConditionFalse
+		newCondition.Reason = string(ReasonWorkflowFailed)
+		newCondition.Message = "Image name is not found in the workflow"
+	} else {
+		build.Status.ImageStatus.Image = image
+	}
```

Comment:
This logic is related to the controller's core logic as it determines the condition from a state. We can keep the `NewSuccessCondition` and `NewFailedCondition` here and use these functions inside the controller after doing the business logic.